event_model/tests/test_em.py

```diff
@@ -340,6 +340,154 @@ def test_document_router_():
     dr('stop', run_bundle.compose_stop(), validate=True)
 
 
+def test_document_router_dispatch_event():
+
+    mutable = []  # used for counting calls
+
+    # example documents
+    event1 = {'data': {'x': 1},
+              'timestamps': {'x': 0.},
+              'uid': 'placeholder X',
+              'descriptor': 'placeholder Y',
+              'time': 0.,
+              'seq_num': 1}
+    event2 = {'data': {'x': 2},
+              'timestamps': {'x': 1.},
+              'uid': 'placeholder X',
+              'descriptor': 'placeholder Y',
+              'time': 1.,
+              'seq_num': 2}
+    event_page = event_model.pack_event_page(event1, event2)
+
+    class DefinesEventNotEventPage(event_model.DocumentRouter):
+        def event(self, doc):
+            # Just a dumb test that check something particular to these example
+            # documents.
+            assert doc['data']['x'] == doc['seq_num']
+            mutable.append(object())
+
+    dr = DefinesEventNotEventPage()
+    # Test that Event is routed to Event.
+    dr('event', event1)
+    assert len(mutable) == 1
+    mutable.clear()
+    # Test that EventPage is unpacked and routed to Event one at a time.
+    dr('event_page', event_page)
+    assert len(mutable) == 2
+    mutable.clear()
+
+    class DefinesEventPageNotEvent(event_model.DocumentRouter):
+        def event_page(self, doc):
+            # Just a dumb test that check something particular to these example
+            # documents.
+            assert doc['data']['x'][0] == 1
+            mutable.append(object())
+
+    dr = DefinesEventPageNotEvent()
+    # Test that Event is packed and routed to EventPage.
+    dr('event', event1)
+    assert len(mutable) == 1
+    mutable.clear()
+    # Test that EventPage is routed to EventPage.
+    dr('event_page', event_page)
+    assert len(mutable) == 1
+    mutable.clear()
+
+    class DefinesEventPageAndEvent(event_model.DocumentRouter):
+        def event(self, doc):
+            # Just a dumb test that check something particular to these example
+            # documents.
+            assert doc['data']['x'] == doc['seq_num']
+            mutable.append(object())
+
+        def event_page(self, doc):
+            # Just a dumb test that check something particular to these example
+            # documents.
+            assert doc['data']['x'][0] == 1
+            mutable.append(object())
+
+    dr = DefinesEventPageAndEvent()
+    # Test that Event is routed to Event.
+    dr('event', event1)
+    assert len(mutable) == 1
+    mutable.clear()
+    # Test that EventPage is routed to EventPage.
+    dr('event_page', event_page)
```

Comment:
Maybe there should be two separate mutable lists? One that gets appended to when event gets called, and another when event_page gets called.  Without two separate lists I don't think we have proven that the correct method was called.